File: backend/services/inflow_service.py
from __future__ import annotations

import logging
from typing import Any

from services.inflow_client import inflow_client
from services.cache_service import inflow_cache
from schemas.inflow import InflowDashboardSummary
import services.derived_frames_service  # noqa: F401 — registers callbacks on import
import services.location_frames_service  # noqa: F401 — registers location callbacks on import

logger = logging.getLogger(__name__)

_PAGE_COUNT = 100  # Max 100 per request per Inflow API
_MAX_PAGES = 100   # Safety limit to prevent infinite pagination loops


async def _fetch_all_pages(endpoint: str, extra_params: dict[str, Any] | None = None) -> dict:
    """Auto-paginate using Inflow API params, with DataFrame caching.

    1. Check cache → return immediately on hit
    2. Acquire per-key lock (prevents thundering herd)
    3. Re-check cache (another coroutine may have populated it)
    4. Fetch all pages from API
    5. Store in cache (only updates if data actually changed)
    6. Return response
    """
    # Fast path: cache hit
    cached = inflow_cache.get(endpoint, extra_params)
    if cached is not None:
        logger.debug("Cache hit for %s", endpoint)
        return cached

    # Acquire per-endpoint lock to prevent duplicate fetches
    lock = await inflow_cache.get_lock_for(endpoint, extra_params)
    async with lock:
        # Double-check after lock acquisition
        cached = inflow_cache.get(endpoint, extra_params)
        if cached is not None:
            logger.debug("Cache hit (post-lock) for %s", endpoint)
            return cached

        # Fetch all pages from the Inflow API
        all_data: list[Any] = []
        skip = 0
        total_count = None

        for _ in range(_MAX_PAGES):
            params: dict[str, Any] = {
                "count": str(_PAGE_COUNT),
                "skip": str(skip),
                "includeCount": "true",
            }
            if extra_params:
                params.update(extra_params)

            raw, list_count = await inflow_client.get_paged(endpoint, params=params)

            if total_count is None and list_count is not None:
                total_count = list_count

            if isinstance(raw, list):
                page_data = raw
            elif isinstance(raw, dict):
                page_data = raw.get("data", raw.get("items", []))
            else:
                page_data = []

            all_data.extend(page_data)

            if len(page_data) != _PAGE_COUNT:
                break
            if total_count is not None and len(all_data) >= total_count:
                break
            skip += len(page_data)

        final_count = total_count if total_count is not None else len(all_data)

        # Store in DataFrame cache, then always return from cache
        inflow_cache.put(endpoint, all_data, final_count, extra_params)
        return inflow_cache.get(endpoint, extra_params)


_RECENT_ORDERS_LIMIT = 5


# ── Products ─────────────────────────────────────────────────────────

async def list_products(extra_params: dict[str, Any] | None = None) -> dict:
    return await _fetch_all_pages("products", extra_params)


async def get_product(product_id: int) -> dict:
    return await inflow_client.get(f"products/{product_id}")


async def upsert_product(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("products", json=body)
    inflow_cache.invalidate("products")
    return result


async def get_product_summary(body: dict[str, Any]) -> dict:
    return await inflow_client.post("products/summary", json=body)


# ── Customers ────────────────────────────────────────────────────────

async def list_customers() -> dict:
    return await _fetch_all_pages("customers")


async def get_customer(customer_id: int) -> dict:
    return await inflow_client.get(f"customers/{customer_id}")


async def upsert_customer(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("customers", json=body)
    inflow_cache.invalidate("customers")
    return result


# ── Vendors ──────────────────────────────────────────────────────────

async def list_vendors() -> dict:
    return await _fetch_all_pages("vendors")


async def get_vendor(vendor_id: int) -> dict:
    return await inflow_client.get(f"vendors/{vendor_id}")


async def upsert_vendor(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("vendors", json=body)
    inflow_cache.invalidate("vendors")
    return result


# ── Sales Orders ─────────────────────────────────────────────────────

async def list_sales_orders(extra_params: dict[str, Any] | None = None) -> dict:
    return await _fetch_all_pages("sales-orders", extra_params)


async def get_sales_order(order_id: str) -> dict:
    return await inflow_client.get(f"sales-orders/{order_id}")


async def upsert_sales_order(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("sales-orders", json=body)
    inflow_cache.invalidate("sales-orders")
    return result


# ── Purchase Orders ──────────────────────────────────────────────────

async def list_purchase_orders() -> dict:
    return await _fetch_all_pages("purchase-orders")


async def get_purchase_order(order_id: int) -> dict:
    return await inflow_client.get(f"purchase-orders/{order_id}")


async def upsert_purchase_order(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("purchase-orders", json=body)
    inflow_cache.invalidate("purchase-orders")
    return result


# ── Locations ────────────────────────────────────────────────────────

async def list_locations() -> dict:
    return await _fetch_all_pages("locations")


async def get_location(location_id: int) -> dict:
    return await inflow_client.get(f"locations/{location_id}")


# ── Categories ───────────────────────────────────────────────────────

async def list_categories() -> dict:
    return await _fetch_all_pages("categories")


async def get_category(category_id: int) -> dict:
    return await inflow_client.get(f"categories/{category_id}")


# ── Stock Adjustments ────────────────────────────────────────────────

async def list_stock_adjustments() -> dict:
    return await _fetch_all_pages("stock-adjustments")


async def get_stock_adjustment(adjustment_id: int) -> dict:
    return await inflow_client.get(f"stock-adjustments/{adjustment_id}")


async def upsert_stock_adjustment(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("stock-adjustments", json=body)
    inflow_cache.invalidate("stock-adjustments")
    return result


# ── Stock Transfers ──────────────────────────────────────────────────

async def list_stock_transfers() -> dict:
    return await _fetch_all_pages("stock-transfers")


async def get_stock_transfer(transfer_id: int) -> dict:
    return await inflow_client.get(f"stock-transfers/{transfer_id}")


async def upsert_stock_transfer(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("stock-transfers", json=body)
    inflow_cache.invalidate("stock-transfers")
    return result


# ── Stock Counts ─────────────────────────────────────────────────────

async def list_stock_counts() -> dict:
    return await _fetch_all_pages("stock-counts")


async def get_stock_count(count_id: int) -> dict:
    return await inflow_client.get(f"stock-counts/{count_id}")


async def upsert_stock_count(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("stock-counts", json=body)
    inflow_cache.invalidate("stock-counts")
    return result


# ── Manufacturing Orders ─────────────────────────────────────────────

async def list_manufacturing_orders() -> dict:
    return await _fetch_all_pages("manufacturing-orders")


async def get_manufacturing_order(order_id: int) -> dict:
    return await inflow_client.get(f"manufacturing-orders/{order_id}")


async def upsert_manufacturing_order(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("manufacturing-orders", json=body)
    inflow_cache.invalidate("manufacturing-orders")
    return result


# ── Reference data ───────────────────────────────────────────────────

async def list_currencies() -> dict:
    return await _fetch_all_pages("currencies")


async def list_tax_codes() -> dict:
    return await _fetch_all_pages("tax-codes")


async def list_payment_terms() -> dict:
    return await _fetch_all_pages("payment-terms")


async def list_pricing_schemes() -> dict:
    return await _fetch_all_pages("pricing-schemes")


async def list_adjustment_reasons() -> dict:
    return await _fetch_all_pages("adjustment-reasons")


async def list_operation_types() -> dict:
    return await _fetch_all_pages("operation-types")


# ── Team / Stockroom Users ───────────────────────────────────────────

async def list_team_members() -> dict:
    return await _fetch_all_pages("team-members")


async def list_stockroom_users() -> dict:
    return await _fetch_all_pages("stockroom-users")


# ── Webhooks ─────────────────────────────────────────────────────────

async def list_webhooks() -> dict:
    return await _fetch_all_pages("webhooks")


async def upsert_webhook(body: dict[str, Any]) -> dict:
    result = await inflow_client.put("webhooks", json=body)
    inflow_cache.invalidate("webhooks")
    return result


async def delete_webhook(webhook_id: int) -> None:
    await inflow_client.delete(f"webhooks/{webhook_id}")
    inflow_cache.invalidate("webhooks")


# ── Custom Fields ────────────────────────────────────────────────────

async def list_custom_fields() -> dict:
    return await _fetch_all_pages("custom-fields")


# ── Dashboard summary ────────────────────────────────────────────────

async def get_dashboard_summary() -> dict:
    """Aggregate high-level stats from several Inflow endpoints.

    All data is served from the DataFrame cache. If a cache entry is
    missing, _fetch_all_pages populates it first (API → DataFrame → cache).
    """
    # Ensure every endpoint is in the cache (no-op if already cached)
    products = await _fetch_all_pages("products")
    customers = await _fetch_all_pages("customers")
    vendors = await _fetch_all_pages("vendors")
    sales = await _fetch_all_pages("sales-orders")
    purchases = await _fetch_all_pages("purchase-orders")

    return InflowDashboardSummary(
        productsCount=products["totalCount"],
        customersCount=customers["totalCount"],
        vendorsCount=vendors["totalCount"],
        salesOrdersCount=sales["totalCount"],
        purchaseOrdersCount=purchases["totalCount"],
        recentSalesOrders=(sales["data"] or [])[:_RECENT_ORDERS_LIMIT],
        recentPurchaseOrders=(purchases["data"] or [])[:_RECENT_ORDERS_LIMIT],
    ).model_dump()

File: frontend/src/app/shared/services/inflow.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { shareReplay, tap } from 'rxjs/operators';
import { ApiService } from './api.service';
import {
  InflowListResponse,
  InflowProduct,
  InflowCustomer,
  InflowVendor,
  InflowSalesOrder,
  InflowPurchaseOrder,
  InflowLocation,
  InflowCategory,
  InflowStockAdjustment,
  InflowStockTransfer,
  InflowStockCount,
  InflowManufacturingOrder,
  InflowDashboardSummary,
  FranchiseLocationCity,
  FranchiseLocationOrderRow,
  FranchiseOrderMatrixRow,
} from '../models/inflow.model';

const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

export interface RoastRow {
  sku: string;
  name: string;
  stock: number;
}

export const ROAST_SKUS: Record<string, string> = {
  IF5127699: 'Light Roast',
  IF5127705: 'Medium Roast',
  IF5127683: 'Dark Roast',
};

interface CacheEntry<T> {
  observable: Observable<T>;
  expiry: number;
}

@Injectable({
  providedIn: 'root',
})
export class InflowService {
  private cache = new Map<string, CacheEntry<unknown>>();

  /** Populated by ProductInventoryComponent during its filter step. */
  readonly roastRows$ = new BehaviorSubject<RoastRow[]>([]);

  setRoastRows(rows: RoastRow[]): void {
    this.roastRows$.next(rows);
  }

  constructor(private api: ApiService) {}

  /**
   * Returns a cached observable for the given key. If no valid cache entry
   * exists, calls `factory` to create a new request, pipes it through
   * `shareReplay(1)`, and stores it with a TTL.
   */
  private cached<T>(key: string, factory: () => Observable<T>): Observable<T> {
    const entry = this.cache.get(key) as CacheEntry<T> | undefined;
    if (entry) {
      if (Date.now() < entry.expiry) {
        return entry.observable;
      }
      this.cache.delete(key);
    }
    const obs = factory().pipe(shareReplay(1));
    this.cache.set(key, { observable: obs, expiry: Date.now() + CACHE_TTL_MS });
    return obs;
  }

  /** Clear all cached data, or a single key. */
  clearCache(key?: string): void {
    if (key) {
      this.cache.delete(key);
    } else {
      this.cache.clear();
    }
  }

  // ── Incremental Update ──────────────────────────────────────────────

  /** Trigger an on-demand incremental update on the backend.
   *  Fetches only records modified after the last cached data point. */
  updateData(): Observable<{ updated: Record<string, number>; totalNewRecords: number }> {
    return this.api.get<{ updated: Record<string, number>; totalNewRecords: number }>(
      '/inflow/update'
    );
  }

  // ── Dashboard ───────────────────────────────────────────────────────

  getDashboardSummary(): Observable<InflowDashboardSummary> {
    return this.cached('dashboard-summary', () =>
      this.api.get<InflowDashboardSummary>('/inflow/dashboard-summary')
    );
  }

  // ── Products ────────────────────────────────────────────────────────

  listProducts(): Observable<InflowListResponse<InflowProduct>> {
    return this.cached('products', () =>
      this.api.get<InflowListResponse<InflowProduct>>('/inflow/products?include=inventoryLines,category,defaultPrice').pipe(
        tap(res => this._buildRoastRows(res.data || []))
      )
    );
  }

  private _buildRoastRows(products: InflowProduct[]): void {
    const rows: RoastRow[] = Object.entries(ROAST_SKUS).map(([sku, name]) => {
      const nameLower = name.toLowerCase();
      const p =
        products.find(x => (x.sku || '').trim() === sku) ??
        products.find(x => (x.name || '').toLowerCase() === nameLower);
      const stock = p ? this._computeQty(p) : 0;
      console.log(`[RoastStock] ${name}: matched=${!!p}, sku=${p?.sku}, qty=${p?.quantityOnHand}, lines=${JSON.stringify(p?.inventoryLines)}, stock=${stock}`);
      return { sku, name, stock };
    });
    this.roastRows$.next(rows);
  }

  private _computeQty(p: InflowProduct): number {
    const lines = p.inventoryLines;
    if (lines && lines.length > 0) {
      const total = lines.reduce((sum, l) => sum + (Number(l.quantityOnHand) || 0), 0);
      if (total > 0) return Math.round(total);
    }
    return Math.round(Number(p.quantityOnHand) || Number(p.totalQuantityOnHand) || 0);
  }

  getRoastStock(): Observable<{ data: { sku: string; name: string; stock: number }[] }> {
    return this.api.get('/inflow/roast-stock');
  }

  getProduct(id: number): Observable<InflowProduct> {
    return this.api.get<InflowProduct>(`/inflow/products/${id}`);
  }

  upsertProduct(body: Partial<InflowProduct>): Observable<InflowProduct> {
    return this.api.put<InflowProduct>('/inflow/products', body);
  }

  getProductSummary(body: Record<string, unknown>): Observable<unknown> {
    return this.api.post<unknown>('/inflow/products/summary', body);
  }

  // ── Customers ───────────────────────────────────────────────────────

  listCustomers(): Observable<InflowListResponse<InflowCustomer>> {
    return this.api.get<InflowListResponse<InflowCustomer>>('/inflow/customers');
  }

  getCustomer(id: number): Observable<InflowCustomer> {
    return this.api.get<InflowCustomer>(`/inflow/customers/${id}`);
  }

  upsertCustomer(body: Partial<InflowCustomer>): Observable<InflowCustomer> {
    return this.api.put<InflowCustomer>('/inflow/customers', body);
  }

  // ── Vendors ─────────────────────────────────────────────────────────

  listVendors(): Observable<InflowListResponse<InflowVendor>> {
    return this.api.get<InflowListResponse<InflowVendor>>('/inflow/vendors');
  }

  getVendor(id: number): Observable<InflowVendor> {
    return this.api.get<InflowVendor>(`/inflow/vendors/${id}`);
  }

  upsertVendor(body: Partial<InflowVendor>): Observable<InflowVendor> {
    return this.api.put<InflowVendor>('/inflow/vendors', body);
  }

  // ── Sales Orders ────────────────────────────────────────────────────

  listSalesOrders(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.api.get<InflowListResponse<InflowSalesOrder>>('/inflow/sales-orders');
  }

  listSalesOrdersWithDetails(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.cached('sales-orders-details', () =>
      this.api.get<InflowListResponse<InflowSalesOrder>>(
        '/inflow/sales-orders?include=lines.product,customer,location'
      )
    );
  }

  getSalesOrder(id: string): Observable<InflowSalesOrder> {
    return this.api.get<InflowSalesOrder>(`/inflow/sales-orders/${id}`);
  }

  upsertSalesOrder(body: Partial<InflowSalesOrder>): Observable<InflowSalesOrder> {
    return this.api.put<InflowSalesOrder>('/inflow/sales-orders', body);
  }

  // ── Purchase Orders ─────────────────────────────────────────────────

  listPurchaseOrders(): Observable<InflowListResponse<InflowPurchaseOrder>> {
    return this.api.get<InflowListResponse<InflowPurchaseOrder>>('/inflow/purchase-orders');
  }

  getPurchaseOrder(id: number): Observable<InflowPurchaseOrder> {
    return this.api.get<InflowPurchaseOrder>(`/inflow/purchase-orders/${id}`);
  }

  upsertPurchaseOrder(body: Partial<InflowPurchaseOrder>): Observable<InflowPurchaseOrder> {
    return this.api.put<InflowPurchaseOrder>('/inflow/purchase-orders', body);
  }

  // ── Locations ───────────────────────────────────────────────────────

  listLocations(): Observable<InflowListResponse<InflowLocation>> {
    return this.api.get<InflowListResponse<InflowLocation>>('/inflow/locations');
  }

  getLocation(id: number): Observable<InflowLocation> {
    return this.api.get<InflowLocation>(`/inflow/locations/${id}`);
  }

  // ── Categories ──────────────────────────────────────────────────────

  listCategories(): Observable<InflowListResponse<InflowCategory>> {
    return this.api.get<InflowListResponse<InflowCategory>>('/inflow/categories');
  }

  getCategory(id: number): Observable<InflowCategory> {
    return this.api.get<InflowCategory>(`/inflow/categories/${id}`);
  }

  // ── Stock Adjustments ───────────────────────────────────────────────

  listStockAdjustments(): Observable<InflowListResponse<InflowStockAdjustment>> {
    return this.api.get<InflowListResponse<InflowStockAdjustment>>('/inflow/stock-adjustments');
  }

  getStockAdjustment(id: number): Observable<InflowStockAdjustment> {
    return this.api.get<InflowStockAdjustment>(`/inflow/stock-adjustments/${id}`);
  }

  upsertStockAdjustment(
    body: Partial<InflowStockAdjustment>
  ): Observable<InflowStockAdjustment> {
    return this.api.put<InflowStockAdjustment>('/inflow/stock-adjustments', body);
  }

  // ── Stock Transfers ─────────────────────────────────────────────────

  listStockTransfers(): Observable<InflowListResponse<InflowStockTransfer>> {
    return this.api.get<InflowListResponse<InflowStockTransfer>>('/inflow/stock-transfers');
  }

  getStockTransfer(id: number): Observable<InflowStockTransfer> {
    return this.api.get<InflowStockTransfer>(`/inflow/stock-transfers/${id}`);
  }

  upsertStockTransfer(body: Partial<InflowStockTransfer>): Observable<InflowStockTransfer> {
    return this.api.put<InflowStockTransfer>('/inflow/stock-transfers', body);
  }

  // ── Stock Counts ────────────────────────────────────────────────────

  listStockCounts(): Observable<InflowListResponse<InflowStockCount>> {
    return this.api.get<InflowListResponse<InflowStockCount>>('/inflow/stock-counts');
  }

  getStockCount(id: number): Observable<InflowStockCount> {
    return this.api.get<InflowStockCount>(`/inflow/stock-counts/${id}`);
  }

  upsertStockCount(body: Partial<InflowStockCount>): Observable<InflowStockCount> {
    return this.api.put<InflowStockCount>('/inflow/stock-counts', body);
  }

  // ── Manufacturing Orders ────────────────────────────────────────────

  listManufacturingOrders(): Observable<InflowListResponse<InflowManufacturingOrder>> {
    return this.api.get<InflowListResponse<InflowManufacturingOrder>>(
      '/inflow/manufacturing-orders'
    );
  }

  getManufacturingOrder(id: number): Observable<InflowManufacturingOrder> {
    return this.api.get<InflowManufacturingOrder>(`/inflow/manufacturing-orders/${id}`);
  }

  upsertManufacturingOrder(
    body: Partial<InflowManufacturingOrder>
  ): Observable<InflowManufacturingOrder> {
    return this.api.put<InflowManufacturingOrder>('/inflow/manufacturing-orders', body);
  }

  // ── Reference data ──────────────────────────────────────────────────

  listCurrencies(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/currencies');
  }

  listTaxCodes(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/tax-codes');
  }

  listPaymentTerms(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/payment-terms');
  }

  listPricingSchemes(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/pricing-schemes');
  }

  listAdjustmentReasons(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/adjustment-reasons');
  }

  listOperationTypes(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/operation-types');
  }

  // ── Team / Stockroom Users ──────────────────────────────────────────

  listTeamMembers(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/team-members');
  }

  listStockroomUsers(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/stockroom-users');
  }

  // ── Webhooks ────────────────────────────────────────────────────────

  listWebhooks(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/webhooks');
  }

  upsertWebhook(body: Record<string, unknown>): Observable<unknown> {
    return this.api.put<unknown>('/inflow/webhooks', body);
  }

  deleteWebhook(id: number): Observable<unknown> {
    return this.api.delete<unknown>(`/inflow/webhooks/${id}`);
  }

  // ── Custom Fields ───────────────────────────────────────────────────

  listCustomFields(): Observable<InflowListResponse> {
    return this.api.get<InflowListResponse>('/inflow/custom-fields');
  }

  // ── Derived Order Frames ──────────────────────────────────────────

  listNaFranchiseOrders(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.cached('na-franchise-orders', () =>
      this.api.get<InflowListResponse<InflowSalesOrder>>('/inflow/na-franchise-orders')
    );
  }

  listFranchiseStoreOrders(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.cached('franchise-store-orders', () =>
      this.api.get<InflowListResponse<InflowSalesOrder>>('/inflow/franchise-store-orders')
    );
  }

  listOnlineOrders(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.cached('online-orders', () =>
      this.api.get<InflowListResponse<InflowSalesOrder>>('/inflow/online-orders')
    );
  }

  // ── Franchise Location Frames ────────────────────────────────────

  listFranchiseLocationCities(): Observable<InflowListResponse<FranchiseLocationCity>> {
    return this.cached('franchise-location-cities', () =>
      this.api.get<InflowListResponse<FranchiseLocationCity>>(
        '/inflow/franchise-location-orders/cities'
      )
    );
  }

  getFranchiseLocationOrders(citySlug: string): Observable<InflowListResponse<FranchiseLocationOrderRow>> {
    return this.cached(`franchise-location-orders:${citySlug}`, () =>
      this.api.get<InflowListResponse<FranchiseLocationOrderRow>>(
        `/inflow/franchise-location-orders/${citySlug}`
      )
    );
  }

  // ── Franchise Order Matrix ─────────────────────────────────────

  listFranchiseOrderMatrix(): Observable<InflowListResponse<FranchiseOrderMatrixRow>> {
    return this.cached('franchise-order-matrix', () =>
      this.api.get<InflowListResponse<FranchiseOrderMatrixRow>>('/inflow/franchise-order-matrix')
    );
  }

  // ── Latest Franchise Orders ──────────────────────────────────

  listLatestFranchiseOrders(): Observable<InflowListResponse<FranchiseOrderMatrixRow>> {
    return this.cached('latest-franchise-orders', () =>
      this.api.get<InflowListResponse<FranchiseOrderMatrixRow>>('/inflow/latest-franchise-orders')
    );
  }

  // ── Franchise Product Matrix ─────────────────────────────────

  listFranchiseProductMatrix(): Observable<InflowListResponse<FranchiseOrderMatrixRow>> {
    return this.cached('franchise-product-matrix', () =>
      this.api.get<InflowListResponse<FranchiseOrderMatrixRow>>('/inflow/franchise-product-matrix')
    );
  }

  // ── Latest Franchise Product Orders ──────────────────────────

  listLatestFranchiseProductOrders(): Observable<InflowListResponse<FranchiseOrderMatrixRow>> {
    return this.cached('latest-franchise-product-orders', () =>
      this.api.get<InflowListResponse<FranchiseOrderMatrixRow>>('/inflow/latest-franchise-product-orders')
    );
  }

  // ── Active Orders This Week ───────────────────────────────────

  /** This week's (Sat–Fri) sales orders that have not yet been completed.
   *  Not cached — fetched fresh on each call so fulfillment status stays current. */
  listActiveOrdersWeek(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.api.get<InflowListResponse<InflowSalesOrder>>('/inflow/active-orders-week');
  }

  /** All active (non-completed, non-terminal) sales orders.
   *  Not cached — fetched fresh on each call so fulfillment status stays current. */
  listActiveOrders(): Observable<InflowListResponse<InflowSalesOrder>> {
    return this.api.get<InflowListResponse<InflowSalesOrder>>('/inflow/active-orders');
  }
}
